# FuNTB

*`FuNTB`* is a free, open-source Python tool designed for analyzing output from MTBSeq v.1.0. This standalone command-line tool enables the comparison of Single Nucleotide Polymorphisms (SNPs) across phenotypically diverse sets of Mycobacterium tuberculosis samples. By structuring the data into phenotype-centered networks, FUN-TB allows users to visualize which genes are altered and how these alterations correlate with specific phenotypes. The tool provides insights into gene alterations by representing them as nodes, with node size and edge width reflecting the strength of their relationship to particular phenotypes. 

<p align="center">
  <img src="images/FUN-TB_Van_Gogh_4.jpg" width="50%"/>
</p>

*`FuNTB`* consists of three core scripts:

1. **Variation Dictionary Creation**: Generates a dictionary of variations found in the `MTBSeq` output.
2. **Phenotype-Based Sample List Generation**: Generates a dictionary of variations found in the `MTBSeq` output.
3. **Phenotype-Centric and Gene-surrounded Network Structuring**: Constructs networks based on the relationship between phenotypic traits and gene alterations, making it easy to identify key genetic markers associated with specific phenotypes.

## Variation dictionary creation script
This script aims to create a Python dictionary object from an MTBSeq V.1.0 file. The dictionary summarizes the presence of SNPs, along with the position and frequency parameters of each altered gene per sample.

<p align="center">
  <img src="images/Github (1).PNG"/>
</p>


## Phenotype-based samples lists generation script
This second script aims to generate a series of lists containing sample IDs. These lists are constructed such that samples within the same list share common phenotypical features and clinical values. The input CSV file must follow the format below to ensure that the clinical data is processed correctly.

<p align="center">
  <img src="images/Github (2).PNG"/>
</p>

### Required Columns:
- **Sample_ID**: A unique identifier for each sample.

### Optional Metadata Columns (Can vary depending on the dataset but should be consistent within a file):
- **Age:** Age of the individual (numeric).
- **Gender:** Gender of the individual (e.g., `Male`, `Female`, `Other`).
- **Year_of_sample:** Year when the sample was collected (e.g., `2020`, `2021`, `2022`, `2023`).
- **Sequencing_technology:** The sequencing technology used (e.g., `Illumina`, `Nanopore`, `PacBio`).
- **Geographic_location:** Location where the sample was collected (e.g., `Mexico`, `USA`, `Spain`, `Germany`).
- **Disease_status:** Clinical status of the individual (e.g., `HIV Positive`, `Tuberculosis`, `Healthy`).
- **Treatment:** Treatment the individual is receiving (if applicable, e.g., `Antibiotic`, `Antiretrovirals`, `None`).
- **Other_metadata:** Any additional information relevant to the study.

#### Example Clinical Data File (`sample_clinical_data.csv`):
| Sample_ID | AGE | GENDER | YEAR_OF_SAMPLE | SEQUENCING_TECHNOLOGY | GEOGRAPHIC_LOCATION | DISEASE_STATUS  | TREATMENT       |
|-----------|-----|--------|----------------|-----------------------|---------------------|-----------------|----------------|
| S1        | 45  | Male   | 2021           | Illumina              | USA                 | HIV positive    | Antiretrovirals |
| S2        | 60  | Female | 2019           | Nanopore              | Brazil              | Tuberculosis    | Antibiotics     |
| S3        | 30  | Male   | 2020           | PacBio                | Germany             | Healthy         | None           |
| S4        | 50  | Female | 2022           | Illumina              | Canada              | HIV positive    | Antiretrovirals |

#### File Requirements:
1. The file must be in CSV format.
2. The first column (`Sample_ID`) is mandatory and should contain unique identifiers.
3. Other columns can vary based on the dataset, but all values within a column should be consistent (e.g., no mixing of strings and numeric values in the same column).
4. Ensure there are no extra spaces or typos in column names or values.

#### Flexible processing
The script dynamically adapts to different metadata columns. Any metadata category can be used for grouping and filtering if the column name is correctly provided when running the script.

## Phenotype-centric and gene-surrounded networks structuration script.

The third script takes as input the files generated by the previous two scripts. It begins by extracting and summarizing gene information from samples within each specified list. Subsequently, it conducts comparisons among these groupings to identify positions that exhibit at least one alteration and are absent in the contrasting sets.

The script visualizes the results by representing variations in node size and their positions in the network. The contribution of each phenotype to node size is depicted through edge width. Node color serves as an indicator of group membership, and nodes present in more than one group are visually highlighted.

In summary, the script provides a comprehensive analysis of genetic variations, emphasizing their presence or absence in different phenotypic groups. The resulting network visualization offers insights into the relationships between genetic alterations and phenotypic characteristics.

<p align="center">
  <img src="images/Github (3).PNG"/>
</p>

## Table of Contents

- [Features](#features)
- [Set Up](#Setup)
- [Usage](#usage)
- [Documentation](#documentation)
- [Example](https://colab.research.google.com/drive/1eCRk_MqbV8FBXqYg2M0kppKe5qcT6XvT#scrollTo=s-bzMeQjFMcT)
- [Contributing](#contributing)
- [License](#license)
- [Install](install.md)

## üöÄ Features

- Analyze genetic variations related to SNPs.
- Construct phenotype-centric networks based on genetic data.
- Visualize and hypothesize complex relationships between genes and phenotypes.

## ‚öôÔ∏è Setup

Ensure you have these or similar versions installed on your system. 
- Python: 3.11.5
- Pandas: 2.0.2
- Numpy: 1.24.3
- Matplotlib: 3.7.1
- Networkx: 3.1
- Seaborn: 0.13.0
- Scipy: 1.11.1

## üìå Usage

### Variation dictionary creation script

The script processes three input parameters:
1. The **¬¥MTBSeq¬¥** output file, including its file path.
2. The **clinical data** file in CSV format.
3. The name for the resulting output file.

To run the script, use the following command: 

```bash
python FunTB_dictionary.py MTBseq_file.tab Clinical_Data.csv Dictionary_Output_File_name
```

The script generates a `txt` file summarizing information for each sample. The output follows this structure:
- Sample 1
  - Gene 1
    - Total variations
      - integer representing number of alterations.
    - variation positions
      - variable position: frequency of variation
    - symbolic mutations
      - variation symbol: frequency of appearances
For example:

Sample 1
  Gene 1
    Total Variations
      - 10
    Variation Positions
      - Position 345: 50
      - Position 678: 30
    Symbolic Mutations
      - 'A623T'
      - 'G904A'

### Phenotype-based samples lists generation script

This script requires two input parameters:
1. The clinical data file in CSV format.
2. The name for the resulting output files.

To run the script, use the following command:

```bash
python Sample_Grouping__Creation.py Clinical_Data.csv Column_of_Interest_1 Column_of_Interest_2 Column_of_Interest_3 ... Column_of_Interest_n
```
The script generates a series of `txt` files, each containing the IDs of samples that share common clinical values. THe output format is as follows:

- Group name 1

| sample id   |
|     :---:   |
| sample id 1 |
| sample id 2 |
|      .      |
|      .      |
|      .      |
| sample id n |
|             |

### Phenotype-centric and gene-surrounded networks structuration script

The third script requires a minimum of eight input parameters:
- Network Name (`string`): Defines the output file name for the network.
- Variation Dictionary File (`txt`): Specifies the file containing variation data.
- Number of Pareto's frontiers (`integer`): Determines the number of pareto frontiers will take into account to keep genes in the final network.
- Alpha coefficient (`float`): Weighs the contribution of the Alteration Density Score (ADS), whose optimized value is 0.9385527090157504.
- Beta coefficient (`float`): Weighs the contribution of the Alteration Density Score (DAGS), whose optimized value is 0.0007787658410143285.
- Gamma coefficient (`float`): Weighs the contribution of the Alteration Density Score (CDAS), whose optimized value is 0.060668525143235286.
- Sample Lists Files (`txt`): Contain lists of samples from groups with contrasting phenotypic features.

Each of these scores (ADS, DAGS, and CDAS) contributes to calculating the Comprehensive Alteration Impact Score (CAIS), which determines the node sizes in the network.

To execute this script, run the following command:

```bash
python FuNTB.py Network_name Variation_dictionary_file Percentage_factor alpha_coefficient beta_coefficient gamma_coefficient Group_list_1 Group_list_2 ... Group_list_n
```

The output generated are three XML-Network file formats (GraphML, GEXF, GML) an input format for Cytoscape, a network software where we can visualize and edit our resulting genes' relationships. The output format could be summarized in a table like this:

| Source Node  |  Target Node   | CAIS Score | Edge Color | Node Size | Edge Width |
|    :---:     |     :---:      |  :---:     |    :---:   |   :---:   |    :---:   |
|     acee     |   Sensitive    |    7       |    blue    |     5     |     1.0    |
|     rpoc     |   Resistant    |    3       |    green   |     2     |     0.3    |

Once you import the output file in Cytoscape, you can map the different networks' parameters and get the editable format to set some extra settings like node position and distribution, node group based on desired or similar characteristics.  Finally, save your final network image.

## üë• Contributing

- **MSc Axel Alejandro Ramos Garc√≠a** ([GitHub](https://github.com/AxelRamosGarcia)) ‚Äì Main Developer, Maintainer, Bug Fixes & Optimization
- **PhD Paulina Mayel Mejia Ponce** ‚Äì Computational biology for features generation.
- **PhD Alejandro Santos D√≠az** ‚Äì Algorithms advisor.
- **PhD Juan Emmanuel Martinez Ledesma** ‚Äì Statistics advisor.
- **PhD Nelly S√©lem Mojica** ‚Äì Scores development.
- **PhD Cuaht√©moc Licona Cassani** - Biological interpretation.


## üìú License

*`FUN-TB`* is available under MIT License. See License.txt for more details.
